File: books/seven_concurrencies_in_7_weeks/actors/hello_actors/hello_actors.exs
import :timer, only: [ sleep: 1 ]

# START:talker
defmodule Talker do
# START:loop
  def loop do
    receive do
      {:greet, name} -> IO.puts("Hello #{name}")
      {:praise, name} -> IO.puts("#{name}, you're amazing")
      {:celebrate, name, age} -> IO.puts("Here's to another #{age} years, #{name}")
    end
    loop
  end
# END:loop
end
# END:talker

# START:main
pid = spawn(&Talker.loop/0)
send(pid, {:greet, "Huey"})
send(pid, {:praise, "Dewey"})
send(pid, {:celebrate, "Louie", 16})
sleep(1000)
# END:main

File: books/seven_concurrencies_in_7_weeks/actors_scala/01.HelloActors/src/main/scala/com/paulbutcher/HelloActors.scala
package com.paulbutcher

import akka.actor._

// START:messages
case class Greet(name: String)
case class Praise(name: String)
case class Celebrate(name: String, age: Int)
// END:messages

class Talker extends Actor {
  def receive = {
    case Greet(name) => println(s"Hello $name")
    case Praise(name) => println(s"$name, you're amazing")
    case Celebrate(name, age) => println(s"Here's to another $age years, $name")
  }
}

// START:main
object HelloActors extends App {
  val system = ActorSystem("HelloActors")
  val talker = system.actorOf(Props[Talker], "talker")

  talker ! Greet("Huey")
  talker ! Praise("Dewey")
  talker ! Celebrate("Louie", 16)

  Thread.sleep(1000) // Better solution is needed
  system.terminate()
}
// END:main
